irispy/new_sji.py

```diff
@@ -135,13 +124,13 @@ def read_iris_sji_level2_fits(filename):
     """
 
     # Open a fits file
-    my_file = fits.open(filename)
+    my_file = fits.open(filename, memmap=True, do_not_scale_image_data=True)
     # Derive WCS, data and mask for NDCube from fits file.
     wcs = WCS(my_file[0].header)
     data = my_file[0].data
     data_nan_masked = my_file[0].data
-    data_nan_masked[data == BAD_PIXEL_VALUE] = np.nan
-    mask = data_nan_masked == BAD_PIXEL_VALUE
+    data_nan_masked[data == BAD_PIXEL_VALUE_UNSCALED] = BAD_PIXEL_VALUE
+    mask = data_nan_masked == BAD_PIXEL_VALUE_UNSCALED
```

Comment:
There should be an if statement here checking whether ```memmap is True```.  If so, use ```BAD_PIXEL_VALUE_UNSCALED```, else use ```BAD_PIXEL_VALUE```.

```BAD_PIXEL_VALUE_UNSCALED``` also needs to be defined at the top of the file underneath where ``````BAD_PIXEL_VALUE``` is.

Reply:
I wrote these lines to fix a bug before leaving the office, I will change it next week.

Reply:
Sure we can work through this.  First of all, the mask is not a nan mask.  The mask (```SJICube.mask```) is a boolean array (True/False)  with the same shape as the data array.  ```True``` means the corresponding pixel in the data array (```SJICube.data```) is masked, i.e. is bad data.  ```False``` means the opposite.  Using this boolean mask means whether a pixel is defined as good or bad is independent of its value in the data array.

The next issue is what value the unexposed pixels should have.  In truth, this doesn't actually matter because these pixels are masked.  All that matters is that we can identify them when we are creating the mask.  In the FITS files, the unexposed pixels have a value of ```-200``` (assuming data is scaled).  When I originally wrote the ``IRISSpectrograph``` class, I (perhaps unwisely) decided to change those values to NaN.  For the time being we can copy that convention in ```SJICube```.

So, finally to the specifics of the code.  Following the above convention, we want to set the data value of unexposed pixels to ```nan``` and set the mask values of those pixels to ```True```.  Previously this was simply done by finding which pixels have data values of ```-200```.  However, if memory mapping is turned on, unexposed pixels will no longer have a value of ```-200```, but some other value, let's call it ```BAD_PIXEL_VALUE_UNSCALED```.  So the change to the above code will be something like:
```
# Here I assume I've renamed BAD_PIXEL_VALUE to BAD_PIXEL_VALUE_SCALED at top of file.
if memmap is True:
    bad_pixel_value= BAD_PIXEL_VALUE_UNSCALED
elif memmap is False:
    bad_pixel_value = BAD_PIXEL_VALUE_SCALED
data_nan_masked[data == bad_pixel_value] = np.nan
mask = data_nan_masked == bad_pixel_value
```
Defining ```BAD_PIXEL_VALUE_UNSCALED``` must also be done.  But I've left other comments on this PR asking Tiago to confirm whether the scaling is always the same for SJI.  Could you ask him that question in person, as I'm not sure he gets these notifications?

Reply:
Thanks, I have talked with him earlier and both of you helped me to have a better understanding of that. One problem on what you wrote is that if memmap is activated, we can't use `np.nan` "because it's not an integer". That why I modify it to 0 for now, hoping a better solving...

I go ask Tiago now.

Reply:
That's a good point @BaptistePellorceAstro.  I suggest we don't in fact change the data values but only form the mask as part of this code snippet.  To be consistent, the same policy should be implemented when ```memmap is False``` and also in ```read_iris_spectrograph_level2_fits```.